File: Data Structures & Algorithms/rotting-fruit/submission-0.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    orangesRotting(grid) {
        if (grid.length === 0) {
            return 0;
        }

        const rows = grid.length;
        const cols = grid[0].length;
        const queue = [];
        let freshOranges = 0;
        let minutes = 0;

        // Init queue with rotten oranges and count fresh oranges
        for (let i = 0; i < rows; i++) {
            for (let j = 0; j < cols; j++) {
                if (grid[i][j] === 2) {
                    queue.push([i, j]);
                } else if (grid[i][j] === 1) {
                    freshOranges++;
                }
            }
        }

        if (freshOranges === 0) {
            return 0;
        }

        const directions = [
            [0, 1], [0, -1], [1, 0], [-1, 0]
        ];

        while (queue.length > 0) {
            let size = queue.length;

            for (let i = 0; i < size; i++) {
                const [row, col] = queue.shift();

                // Check adjacent cells
                for (const [dr, dc] of directions) {
                    const newRow = row + dr;
                    const newCol = col + dc;

                    if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && grid[newRow][newCol] === 1) {
                        grid[newRow][newCol] = 2; // rot the orange
                        freshOranges--;
                        queue.push([newRow, newCol]); // add rotten orange to the shit
                    }
                }
            }
            if (queue.length > 0) {
                minutes++;
            }
        }
        if (freshOranges > 0) {
            return -1;
        }
        return minutes;
    }
}

File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if not heights:
            return []
        
        rows, cols = len(heights), len(heights[0])
        atlantic_reachable = set()
        pacific_reachable = set()

        def dfs(row, col, reachable):
            if (row, col) in reachable:
                return
            reachable.add((row, col))

            directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
            for dr, dc in directions:
                new_row, new_col = row + dr, col + dc
                if 0 <= new_row < rows and 0 <= new_col < cols and heights[new_row][new_col] >= heights[row][col]:
                    dfs(new_row, new_col, reachable)
        
        for col in range(cols):
            dfs(rows - 1,  col, atlantic_reachable)
        for row in range(rows):
            dfs(row, cols - 1, atlantic_reachable)
        
        for col in range(cols):
            dfs(0, col, pacific_reachable)
        for row in range(rows):
            dfs(row, 0, pacific_reachable)
        
        result = []
        for row in range(rows):
            for col in range(cols):
                if (row, col) in pacific_reachable and (row, col) in atlantic_reachable:
                    result.append([row, col])
        return result